heartbeat/monitors/resolver.go

```diff
@@ -0,0 +1,102 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package monitors
+
+import (
+	"fmt"
+	"net"
+)
+
+// Resolver lets us define custom DNS resolvers similar to what the go stdlib provides, but
+// potentially with custom functionality
+type Resolver interface {
+	// ResolveIPAddr is an analog of net.ResolveIPAddr
+	ResolveIPAddr(network string, host string) (*net.IPAddr, error)
+	// LookupIP is an analog of net.LookupIP
+	LookupIP(host string) ([]net.IP, error)
+}
+
+// stdResolver uses the go std library to perform DNS resolution.
+type stdResolver struct{}
+
+// We only ever need one instance of this
+var StdResolver = stdResolver{}
```

Comment:
I'm not a fan of such defined globals/singletons. Is there a way to create an instance once it's needed?

Reply:
Singletons and globals do indeed create problems with using unnecessary memory, data races with concurrent access etc, testability as well. 

Since we exclusively use the `Resolver` interface we're OK on testability. Since we don't have any state (it's an empty struct), and hence to threadsafety issues there's no worries around any of the singleton concerns that I can think of. If there are other concerns, please let me know!

Contrast that with the `StaticResolver` type below this one which does have those concerns. For that one there is no singleton instance, but rather a constructor is used.

Reply:
>  If there are other concerns, please let me know!

Well, in general this is an anti-pattern. You won't be able to detect if any other piece of code modifies it, including unit tests.

Reply:
There are exceptions to every rule. This would be an exception because none of the reasons why singletons are generally an anti-pattern apply in this specific case. If there are specific drawbacks here, please let me know.

Reply:
I'm happy to hear other voices here. I admit I didn't spend any time on heartbeat implementation.

Reply:
BTW, WRT modifying it, go won't let you because the var is typed as `stdResolver{}`. You could assign it a different instance of `stdResolver`, but there's no reason anyone would do that, and even if you did, it wouldn't do anything different in terms of behavior, since it's just an empty struct.

Reply:
I'd also prefer not to use `var` for globals, even if one can not really assign it. Singleton pattern might be ok, as this is an empty struct, but still prefer to provide a constructor function like so:

```
var stdResolverSingleton Resolver = stdResolver{}

func (New)StandardResolver() Resolver {
  return stdResolverSingleton
}
```

As the struct is empty the singleton doesn't even save you an alloc. It's a very small micro-optimisation saving you some call into the runtime converting the struct to an interface.

Reply:
Changed, given that there's consensus around using a constructor I just ditched the var.